File: opcua-webservice-spring/examples/uawebservicetest/src/main/java/org/opcfoundation/uawebservicetest/testdigitaltwin/SubmodelTestType.java
package org.opcfoundation.uawebservicetest.testdigitaltwin;

import org.eclipse.milo.opcua.sdk.core.ValueRank;
import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.*;
import org.eclipse.milo.opcua.stack.core.types.structured.Argument;
import org.eclipse.milo.opcua.stack.core.types.structured.EUInformation;
import org.eclipse.milo.opcua.stack.core.types.structured.Range;
import org.opcfoundation.webserver.addressspace.nodes.UaMethod;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.addressspace.nodes.UaVariable;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaDataTypes;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaVariableTypes;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.submodel.SubmodelType;
import org.opcfoundation.webserver.service.message.digitaltwin.*;
import org.opcfoundation.webserver.types.common.UaStructureUtilities;

import java.time.OffsetDateTime;
import java.util.*;
import java.util.concurrent.CompletableFuture;

public class SubmodelTestType extends SubmodelType {
    private final UaVariable Bool;
    private final UaVariable Double;
    private final UaMethod   Method;

    private final UaObject ElementCollectionTestA;
    private final UaObject ElementCollectionTestB;
    private final UaObject Reference;
    private final UaObject ElementList;
    private final UaObject EventElement;

    public SubmodelTestType(
            ElementCollectionTestAType elementCollectionTestAType,
            ElementCollectionTestBType elementCollectionTestBType,
            ReferenceElementTestType referenceElementTestType,
            ElementListTestType elementListTestType,
            EventElementTestType eventElementTestType,
            DigitalTwinSpace space)
    {
        super("SubmodelTestType", new LocalizedText("SubmodelTestType"), space);

        Bool = addPropertyElement(
                "Bool",
                new LocalizedText("Bool"),
                new LocalizedText("Bool member"),
                UaDataTypes.Boolean,
                true);

        Double = addPropertyElement(
                "Double",
                new LocalizedText("Double"),
                new LocalizedText("Double member"),
                UaDataTypes.Double,
                true,
                true,
                null,
                UaVariableTypes.BaseAnalogType,
                false);

        addSubElementOfProperty(
                Double,
                "EURange",
                UaStructureUtilities.toVariant(new Range(0.0,50.0)));

        addSubElementOfProperty(
                Double,
                "EngineeringUnits",
                UaStructureUtilities.toVariant(new EUInformation(null,1, new LocalizedText("C"), new LocalizedText("Temperature C"))));

        List<Argument> inputArguments = new ArrayList<>();
        inputArguments.add(new Argument("In", NodeIds.String, ValueRank.Scalar.getValue(), null, LocalizedText.NULL_VALUE));

        List<Argument> outArguments = new ArrayList<>();
        outArguments.add(new Argument("Out", NodeIds.String, ValueRank.Scalar.getValue(), null, LocalizedText.NULL_VALUE));

        Method = addOperationElement(
                "Method",
                new LocalizedText("Method"),
                new LocalizedText("Test method interface"),
                inputArguments,
                outArguments,
                false);

        ElementCollectionTestA = addElementCollection(
                elementCollectionTestAType,
                "CollectionA",
                new LocalizedText("CollectionA"),
                new LocalizedText("CollectionA member"),
                true);

        ElementCollectionTestB = addElementCollection(
                elementCollectionTestBType,
                "CollectionB",
                new LocalizedText("CollectionB"),
                new LocalizedText("CollectionB member"),
                true);

        Reference = addReferenceElement(
                referenceElementTestType,
                "Reference",
                new LocalizedText("Reference"),
                new LocalizedText("Reference member"),
                false);

        ElementList = addElementList(
                elementListTestType,
                "ElementList",
                new LocalizedText("ElementList"),
                new LocalizedText("Element list member"),
                false);

        EventElement = addEventElement(
                eventElementTestType,
                "EventElement",
                new LocalizedText("EventElement"),
                new LocalizedText("EventElement member"),
                true);
    }

    public CompletableFuture<ReadPropertyValuesResponse> onReadPropertyValues(ReadPropertyValuesRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            ReadPropertyValuesResponse response = new ReadPropertyValuesResponse();

            for (String item: request.getPropertyNames())
            {
                if (item.equals(Bool.browseName()))
                {
                    response.setValue(item, new Variant(true));
                } else if (item.equals(Double.browseName())) {
                    response.setValue(item, new Variant(25.5));
                }
            }

            return response;
        });
    }

    @Override
    public CompletableFuture<WritePropertyValuesResponse> onWritePropertyValues(WritePropertyValuesRequest request) {
        return CompletableFuture.supplyAsync(() -> {
            System.out.println("Write object, " + request.getId());

            WritePropertyValuesResponse response = new WritePropertyValuesResponse();

            for (Map.Entry<String, Variant> item: request.getPropertyNamesAndValues().entrySet())
            {
                System.out.println(item);
                response.setWriteValueResult(item.getKey(), StatusCode.GOOD);
            }

            return response;
        });
    }

    @Override
    public CompletableFuture<InvokeOperationResponse> onInvokeOperation(InvokeOperationRequest request) {
        return CompletableFuture.supplyAsync(()->{
            List<Variant> outputArguments = new ArrayList<>();

            if (Method.browseName().equals(request.getOperationName()))
            {
                System.out.println("Object id: " + request.getId());
                System.out.println("Method call: " + request.getOperationName());

                String output = "Object " + request.getId();
                output += " receive value ";

                if (!request.getInputArguments().isEmpty())
                {
                    output += request.getInputArguments().get(0).getValue();
                }

                outputArguments.add(Variant.ofString(output));
            }

            return new InvokeOperationResponse(outputArguments);
        });
    }

    @Override
    public CompletableFuture<GetElementsResponse> onGetElements(GetElementsRequest request) {
        GetElementsResponse response = new GetElementsResponse();

        if (request.getId().equals("1")) {
            response.add(Method.browseName());
            response.add(Reference.browseName());
        } else if (request.getId().equals("2")) {
            response.add(Method.browseName());
        } else {
            response.add(Double.browseName());
            response.add(Method.browseName());
            response.add(Reference.browseName());
            response.add(ElementList.browseName());
        }

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public CompletableFuture<GetDescriptorResponse> onGetDescriptor(GetDescriptorRequest request) {
        return CompletableFuture.supplyAsync(() -> {
            return new GetDescriptorResponse(
                    new LocalizedText("Submodel-" + request.getId()),
                    new LocalizedText("Submodel with id " + request.getId()));
        });
    }

    @Override
    public CompletableFuture<ReadPropertyHistoryValuesResponse> onReadPropertyHistoryValues(ReadPropertyHistoryValuesRequest request) {
        ReadPropertyHistoryValuesResponse response = new ReadPropertyHistoryValuesResponse();
        DateTime now = DateTime.now();

        if (null != request.getReadRawDetails())
        {
            System.out.println(request.getReadRawDetails().toString());
        } else if (null != request.getReadAtTimeDetails()) {
            System.out.println(request.getReadAtTimeDetails().toString());
        } else if (null != request.getReadProcessedDetails()) {
            System.out.println(request.getReadProcessedDetails().toString());
        }

        for (int i=0; i<10; ++i)
        {
            DataValue value = new DataValue(
                    Variant.ofInt32(i),
                    StatusCode.GOOD,
                    now);

            response.addDataValue(value);
        }

        return CompletableFuture.completedFuture(response);
    }
}

File: opcua-webservice-node/examples/testserver/src/testdigitaltwin/SubmodelTestType.ts
import { StatusCodes } from "opcua-webapi";
import {
    DataTypeIds,
    makeUaStatusCode,
    UaArgument,
    UaDataValue,
    UaEUInformation,
    UaLocalizedText,
    UaNodeId,
    UaRange,
    UaValueRank,
    UaVariant,
    UaVariantType,
} from "opcua-webapi-ts";
import {
    UaDataTypes,
    UaMethod,
    UaModellingRule,
    UaObject,
    UaVariable,
    UaVariableTypes,
} from "opcua-webservice-node";
import { DigitalTwinSpace,SubmodelType } from "opcua-webservice-node";
import {
    GetDescriptorRequest,
    GetDescriptorResponse,
    GetElementsRequest,
    GetElementsResponse,
    InvokeOperationRequest,
    InvokeOperationResponse,
    ReadPropertyHistoryValuesRequest,
    ReadPropertyHistoryValuesResponse,
    ReadPropertyValuesRequest,
    ReadPropertyValuesResponse,
    WritePropertyValuesRequest,
    WritePropertyValuesResponse,
} from "opcua-webservice-node";
import { ElementCollectionTestAType } from "./ElementCollectionTestAType";
import { ElementCollectionTestBType } from "./ElementCollectionTestBType";
import { ElementListTestType } from "./ElementListTestType";
import { EventElementTestType } from "./EventElementTestType";
import { ReferenceElementTestType } from "./ReferenceElementTestType";

export class SubmodelTestType extends SubmodelType {
    private readonly boolElement: UaVariable;
    private readonly doubleElement: UaVariable;
    private readonly methodElement: UaMethod;
    private readonly elementCollectionTestA: UaObject;
    private readonly elementCollectionTestB: UaObject;
    private readonly referenceElement: UaObject;
    private readonly elementList: UaObject;
    private readonly eventElement: UaObject;

    constructor(
        elementCollectionTestAType: ElementCollectionTestAType,
        elementCollectionTestBType: ElementCollectionTestBType,
        referenceElementTestType: ReferenceElementTestType,
        elementListTestType: ElementListTestType,
        eventElementTestType: EventElementTestType,
        space: DigitalTwinSpace,
    ) {
        super("SubmodelTestType", new UaLocalizedText("SubmodelTestType"), space);
        this.description = new UaLocalizedText("SubmodelTestType");

        this.boolElement = this.addPropertyElement(
            "Bool",
            new UaLocalizedText("Bool"),
            new UaLocalizedText("Bool member"),
            UaDataTypes.Boolean,
            true,
        );

        this.doubleElement = this.addPropertyElement(
            "Double",
            new UaLocalizedText("Double"),
            new UaLocalizedText("Double member"),
            UaDataTypes.Double,
            true,
            false,            
            -1,
            false,
            UaVariableTypes.BaseAnalogItemType,
        );

        this.addSubElementOfProperty(
            this.doubleElement,
            "EURange",
            UaVariant.extensionObject(new UaRange(10.0, 50.0).toExtensionObject()),
        );

        this.addSubElementOfProperty(
            this.doubleElement,
            "EngineeringUnits",
            UaVariant.extensionObject(
                new UaEUInformation(
                    1,
                    new UaLocalizedText("C"),
                    new UaLocalizedText("Temperature C"),
                    null,
                ).toExtensionObject(),
            ),
        );

        const inputArguments = [
            new UaArgument(
                "In",
                UaNodeId.from(DataTypeIds.String),
                UaValueRank.Scalar,
                UaLocalizedText.nullText,
            ),
        ];

        const outputArguments = [
            new UaArgument(
                "Out",
                UaNodeId.from(DataTypeIds.String),
                UaValueRank.Scalar,
                UaLocalizedText.nullText,
            ),
        ];

        this.methodElement = this.addOperationElement(
            "Method",
            new UaLocalizedText("Method"),
            new UaLocalizedText("Test method interface"),
            inputArguments,
            outputArguments,
            false,
        );

        this.elementCollectionTestA = this.addElementCollection(
            elementCollectionTestAType,
            "CollectionA",
            new UaLocalizedText("CollectionA"),
            new UaLocalizedText("CollectionA member"),
            true,
        );

        this.elementCollectionTestB = this.addElementCollection(
            elementCollectionTestBType,
            "CollectionB",
            new UaLocalizedText("CollectionB"),
            new UaLocalizedText("CollectionB member"),
            true,
        );

        this.referenceElement = this.addReferenceElement(
            referenceElementTestType,
            "Reference",
            new UaLocalizedText("Reference"),
            new UaLocalizedText("Reference member"),
            false,
        );

        this.elementList = this.addElementList(
            elementListTestType,
            "ElementList",
            new UaLocalizedText("ElementList"),
            new UaLocalizedText("Element list member"),
            false,
        );

        this.eventElement = this.addEventElement(
            eventElementTestType,
            "EventElement",
            new UaLocalizedText("EventElement"),
            new UaLocalizedText("EventElement member"),
            true,
        );
    }

    override async onReadPropertyValues(request: ReadPropertyValuesRequest): Promise<ReadPropertyValuesResponse> {
        const response = new ReadPropertyValuesResponse();
        response.setValue(this.boolElement.name, UaVariant.boolean(true));
        response.setValue(this.doubleElement.name, UaVariant.double(25.5));
        return response;
    }

    override async onWritePropertyValues(request: WritePropertyValuesRequest): Promise<WritePropertyValuesResponse> {
        console.log(`Write object, ${request.id}`);

        const response = new WritePropertyValuesResponse();
        for (const [key, value] of request.propertyNamesAndValues) {
            console.log(key, value.value);
            response.setWriteValueResult(key, makeUaStatusCode(StatusCodes.Good));
        }

        return response;
    }

    override async onInvokeOperation(request: InvokeOperationRequest): Promise<InvokeOperationResponse> {
        const outputArguments: Array<UaVariant> = [];

        if (this.methodElement.browseName === request.operationName) {
            console.log(`Object id: ${request.id}`);
            console.log(`Method call: ${request.operationName}`);

            let output = `Object ${request.id} receive value `;
            if (request.inputArguments.length > 0) {
                output += String(request.inputArguments[0].value);
            }

            outputArguments.push(UaVariant.string(output));
        }

        return new InvokeOperationResponse(outputArguments);
    }

    override async onGetElements(request: GetElementsRequest): Promise<GetElementsResponse> {
        const response = new GetElementsResponse();
        response.add(this.boolElement.browseName);
        response.add(this.elementCollectionTestA.browseName);
        response.add(this.elementCollectionTestB.browseName);
        response.add(this.eventElement.browseName);

        if (request.id === "1") {
            response.add(this.referenceElement.browseName);
        } else if (request.id === "2") {
            response.add(this.methodElement.browseName);
        } else {
            response.add(this.doubleElement.browseName);
            response.add(this.methodElement.browseName);
            response.add(this.referenceElement.browseName);
            response.add(this.elementList.browseName);
        }

        return response;
    }

    override async onGetDescriptor(request: GetDescriptorRequest): Promise<GetDescriptorResponse> {
        return new GetDescriptorResponse(
            new UaLocalizedText(`Submodel-${request.id}`),
            new UaLocalizedText(`Submodel with id ${request.id}`),
        );
    }

    override async onReadPropertyHistoryValues(request: ReadPropertyHistoryValuesRequest): Promise<ReadPropertyHistoryValuesResponse> {
        const response = new ReadPropertyHistoryValuesResponse();
        let time = new Date();

        if (request.readRawDetails !== null) {
            console.log(request.readRawDetails.toStruct());
        } else if (request.readAtTimeDetails !== null) {
            console.log(request.readAtTimeDetails.toStruct());
        } else if (request.readProcessedDetails !== null) {
            console.log(request.readProcessedDetails.toStruct());
        }

        for (let index = 0; index < 20; ++index) {
            time = new Date(time.getTime() + 5000);
            response.addDataValue(
                new UaDataValue(
                    UaVariant.double(index),
                    makeUaStatusCode(StatusCodes.Good),
                    time,
                ),
            );
        }

        return response;
    }
}